File: go/test/order_entity_test.go
package sdktest

import (
	"encoding/json"
	"os"
	"path/filepath"
	"runtime"
	"strings"
	"testing"
	"time"

	sdk "github.com/voxgig-sdk/abdo-store-sdk"
	"github.com/voxgig-sdk/abdo-store-sdk/core"

	vs "github.com/voxgig/struct"
)

func TestOrderEntity(t *testing.T) {
	t.Run("instance", func(t *testing.T) {
		testsdk := sdk.TestSDK(nil, nil)
		ent := testsdk.Order(nil)
		if ent == nil {
			t.Fatal("expected non-nil OrderEntity")
		}
	})

	t.Run("basic", func(t *testing.T) {
		setup := orderBasicSetup(nil)
		// Per-op sdk-test-control.json skip — basic test exercises a flow
		// with multiple ops; skipping any op skips the whole flow.
		_mode := "unit"
		if setup.live {
			_mode = "live"
		}
		for _, _op := range []string{"create", "load"} {
			if _shouldSkip, _reason := isControlSkipped("entityOp", "order." + _op, _mode); _shouldSkip {
				if _reason == "" {
					_reason = "skipped via sdk-test-control.json"
				}
				t.Skip(_reason)
				return
			}
		}
		// The basic flow consumes synthetic IDs from the fixture. In live mode
		// without an *_ENTID env override, those IDs hit the live API and 4xx.
		if setup.syntheticOnly {
			t.Skip("live entity test uses synthetic IDs from fixture — set ABDOSTORE_TEST_ORDER_ENTID JSON to run live")
			return
		}
		client := setup.client

		// CREATE
		orderRef01Ent := client.Order(nil)
		orderRef01Data := core.ToMapAny(vs.GetProp(
			vs.GetPath([]any{"new", "order"}, setup.data), "order_ref01"))

		orderRef01DataResult, err := orderRef01Ent.Create(orderRef01Data, nil)
		if err != nil {
			t.Fatalf("create failed: %v", err)
		}
		orderRef01Data = core.ToMapAny(orderRef01DataResult)
		if orderRef01Data == nil {
			t.Fatal("expected create result to be a map")
		}

		// LOAD
		orderRef01MatchDt0 := map[string]any{}
		orderRef01DataDt0Loaded, err := orderRef01Ent.Load(orderRef01MatchDt0, nil)
		if err != nil {
			t.Fatalf("load failed: %v", err)
		}
		if orderRef01DataDt0Loaded == nil {
			t.Fatal("expected load result to be non-nil")
		}

	})
}

func orderBasicSetup(extra map[string]any) *entityTestSetup {
	loadEnvLocal()

	_, filename, _, _ := runtime.Caller(0)
	dir := filepath.Dir(filename)

	entityDataFile := filepath.Join(dir, "..", "..", ".sdk", "test", "entity", "order", "OrderTestData.json")

	entityDataSource, err := os.ReadFile(entityDataFile)
	if err != nil {
		panic("failed to read order test data: " + err.Error())
	}

	var entityData map[string]any
	if err := json.Unmarshal(entityDataSource, &entityData); err != nil {
		panic("failed to parse order test data: " + err.Error())
	}

	options := map[string]any{}
	options["entity"] = entityData["existing"]

	client := sdk.TestSDK(options, extra)

	// Generate idmap via transform, matching TS pattern.
	idmap := vs.Transform(
		[]any{"order01", "order02", "order03"},
		map[string]any{
			"`$PACK`": []any{"", map[string]any{
				"`$KEY`": "`$COPY`",
				"`$VAL`": []any{"`$FORMAT`", "upper", "`$COPY`"},
			}},
		},
	)

	// Detect ENTID env override before envOverride consumes it. When live
	// mode is on without a real override, the basic test runs against synthetic
	// IDs from the fixture and 4xx's. Surface this so the test can skip.
	entidEnvRaw := os.Getenv("ABDOSTORE_TEST_ORDER_ENTID")
	idmapOverridden := entidEnvRaw != "" && strings.HasPrefix(strings.TrimSpace(entidEnvRaw), "{")

	env := envOverride(map[string]any{
		"ABDOSTORE_TEST_ORDER_ENTID": idmap,
		"ABDOSTORE_TEST_LIVE":      "FALSE",
		"ABDOSTORE_TEST_EXPLAIN":   "FALSE",
		"ABDOSTORE_APIKEY":         "NONE",
	})

	idmapResolved := core.ToMapAny(env["ABDOSTORE_TEST_ORDER_ENTID"])
	if idmapResolved == nil {
		idmapResolved = core.ToMapAny(idmap)
	}

	if env["ABDOSTORE_TEST_LIVE"] == "TRUE" {
		mergedOpts := vs.Merge([]any{
			map[string]any{
				"apikey": env["ABDOSTORE_APIKEY"],
			},
			extra,
		})
		client = sdk.NewAbdoStoreSDK(core.ToMapAny(mergedOpts))
	}

	live := env["ABDOSTORE_TEST_LIVE"] == "TRUE"
	return &entityTestSetup{
		client:        client,
		data:          entityData,
		idmap:         idmapResolved,
		env:           env,
		explain:       env["ABDOSTORE_TEST_EXPLAIN"] == "TRUE",
		live:          live,
		syntheticOnly: live && !idmapOverridden,
		now:           time.Now().UnixMilli(),
	}
}

File: rb/test/order_entity_test.rb
# Order entity test

require "minitest/autorun"
require "json"
require_relative "../AbdoStore_sdk"
require_relative "runner"

class OrderEntityTest < Minitest::Test
  def test_create_instance
    testsdk = AbdoStoreSDK.test(nil, nil)
    ent = testsdk.Order(nil)
    assert !ent.nil?
  end

  def test_basic_flow
    setup = order_basic_setup(nil)
    # Per-op sdk-test-control.json skip.
    _live = setup[:live] || false
    ["create", "load"].each do |_op|
      _should_skip, _reason = Runner.is_control_skipped("entityOp", "order." + _op, _live ? "live" : "unit")
      if _should_skip
        skip(_reason || "skipped via sdk-test-control.json")
        return
      end
    end
    # The basic flow consumes synthetic IDs from the fixture. In live mode
    # without an *_ENTID env override, those IDs hit the live API and 4xx.
    if setup[:synthetic_only]
      skip "live entity test uses synthetic IDs from fixture — set ABDOSTORE_TEST_ORDER_ENTID JSON to run live"
      return
    end
    client = setup[:client]

    # CREATE
    order_ref01_ent = client.Order(nil)
    order_ref01_data = Helpers.to_map(Vs.getprop(
      Vs.getpath(setup[:data], "new.order"), "order_ref01"))

    order_ref01_data_result = order_ref01_ent.create(order_ref01_data, nil)
    order_ref01_data = Helpers.to_map(order_ref01_data_result)
    assert !order_ref01_data.nil?

    # LOAD
    order_ref01_match_dt0 = {}
    order_ref01_data_dt0_loaded = order_ref01_ent.load(order_ref01_match_dt0, nil)
    assert !order_ref01_data_dt0_loaded.nil?

  end
end

def order_basic_setup(extra)
  Runner.load_env_local

  entity_data_file = File.join(__dir__, "..", "..", ".sdk", "test", "entity", "order", "OrderTestData.json")
  entity_data_source = File.read(entity_data_file)
  entity_data = JSON.parse(entity_data_source)

  options = {}
  options["entity"] = entity_data["existing"]

  client = AbdoStoreSDK.test(options, extra)

  # Generate idmap via transform.
  idmap = Vs.transform(
    ["order01", "order02", "order03"],
    {
      "`$PACK`" => ["", {
        "`$KEY`" => "`$COPY`",
        "`$VAL`" => ["`$FORMAT`", "upper", "`$COPY`"],
      }],
    }
  )

  # Detect ENTID env override before envOverride consumes it. When live
  # mode is on without a real override, the basic test runs against synthetic
  # IDs from the fixture and 4xx's. Surface this so the test can skip.
  entid_env_raw = ENV["ABDOSTORE_TEST_ORDER_ENTID"]
  idmap_overridden = !entid_env_raw.nil? && entid_env_raw.strip.start_with?("{")

  env = Runner.env_override({
    "ABDOSTORE_TEST_ORDER_ENTID" => idmap,
    "ABDOSTORE_TEST_LIVE" => "FALSE",
    "ABDOSTORE_TEST_EXPLAIN" => "FALSE",
    "ABDOSTORE_APIKEY" => "NONE",
  })

  idmap_resolved = Helpers.to_map(
    env["ABDOSTORE_TEST_ORDER_ENTID"])
  if idmap_resolved.nil?
    idmap_resolved = Helpers.to_map(idmap)
  end

  if env["ABDOSTORE_TEST_LIVE"] == "TRUE"
    merged_opts = Vs.merge([
      {
        "apikey" => env["ABDOSTORE_APIKEY"],
      },
      extra || {},
    ])
    client = AbdoStoreSDK.new(Helpers.to_map(merged_opts))
  end

  live = env["ABDOSTORE_TEST_LIVE"] == "TRUE"
  {
    client: client,
    data: entity_data,
    idmap: idmap_resolved,
    env: env,
    explain: env["ABDOSTORE_TEST_EXPLAIN"] == "TRUE",
    live: live,
    synthetic_only: live && !idmap_overridden,
    now: (Time.now.to_f * 1000).to_i,
  }
end
